### PR TITLE
Add LastModifiedInstant as field in Client

### LINKED ISSUE #218: Add last modified field to Client

### PR DESCRIPTION
**Pull request by tankangliang**
## Description

Added new field `LastModifiedInstant` to `Client`. This field is updated every time `Client` is created or edited for now. Adding or deleting notes currently does not update this field.

Fixes #218 

## Testing

Unit tests are done for the field and all its methods.
`LastModifiedInstant` field does not affect equality / hashcode of `Client` since the creation of a new `Client` to check modifies this `LastModifiedInstant` field.

### CODE REVIEW

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 1]
```diff
@@ -143,9 +148,16 @@ public Client toModelType() throws IllegalValueException {
         final ContractExpiryDate modelContractExpiryContractExpiryDate =
                 ParserUtil.parseContractExpiryDate(contractExpiryDate);
 
+        LastModifiedInstant modelLastModifiedInstant;
+        if (lastModifiedInstant == null) {
+            modelLastModifiedInstant = new LastModifiedInstant();
+        } else {
+            modelLastModifiedInstant = new LastModifiedInstant(lastModifiedInstant);
+        }
```
Comment (qwoprocks) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java:159]: When will this be null? Should it just throw an exception when null if its not supposed to be null?

File: src/main/java/seedu/address/model/client/Client.java [hunk 2]
```diff
@@ -31,12 +31,13 @@
     private final ContractExpiryDate contractExpiryDate;
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/client/Client.java:31]: Should the equals and hashCode methods be updated for this class?

File: src/test/java/seedu/address/model/client/LastModifiedInstantTest.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        // missing parts
+        assertFalse(LastModifiedInstant.isValidInstant("-10-10T15:18:35.617617Z")); // missing year
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10T15:18:35.617617Z")); // missing month
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10T15:18:35.617617Z")); // missing day
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-1015:18:35.617617Z")); // missing T
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10T18:35.617617Z")); // missing hour
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10T15:35.617617Z")); // missing minutes
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10T15:18.617617Z")); // missing seconds
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10T15:18:35.617617")); // missing Z
+
```
Comment (qwoprocks) [inline src/test/java/seedu/address/model/client/LastModifiedInstantTest.java:67]: extra newline

File: src/test/java/seedu/address/model/client/LastModifiedInstantTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        assertFalse(LastModifiedInstant.isValidInstant("2020-MM-10T15:18:35.617617Z")); // invalid month
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-DDT15:18:35.617617Z")); // invalid day
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10THH:18:35.617617Z")); // invalid hour
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10T15:MM:35.617617Z")); // invalid minute
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10T15:18:SS.617617Z")); // invalid second
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10T15:18:35.nanosecondZ")); // invalid nanoseconds
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10T15:18:35.617617Y")); // invalid zone
+        assertFalse(LastModifiedInstant.isValidInstant("2020-10-10Z15:18:35.617617Z")); // invalid separator
+        assertFalse(LastModifiedInstant.isValidInstant("2020/10/10T15:18:35.617617Z")); // "/" instead of "-"
+
```
Comment (qwoprocks) [inline src/test/java/seedu/address/model/client/LastModifiedInstantTest.java:79]: extra new line

**Review by qwoprocks — CHANGES_REQUESTED** [review]

File: src/test/java/seedu/address/model/client/LastModifiedInstantTest.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+import java.time.LocalDate;
+
+import org.junit.jupiter.api.Test;
+
+class LastModifiedInstantTest {
+
+    private static final String VALID_INSTANT_1 = "2020-10-10T15:18:35.617617Z";
+    private static final String VALID_INSTANT_2 = "2020-10-11T15:18:35.617617Z";
+
```
Comment (qwoprocks) [inline src/test/java/seedu/address/model/client/LastModifiedInstantTest.java:18]: Can probably add one more test for the empty constructor then check if Instant.now() is < 1s away from the constructed LastModifiedInstant.

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java (hunk 1 above)
Comment (tankangliang) [reply src/main/java/seedu/address/storage/JsonAdaptedClient.java:159]: It can only be null if it doesn't exist in storage, but I feel like the corruption of this data field is too insignificant to have to throw away all existing data on this client.

File: src/main/java/seedu/address/model/client/Client.java (hunk 2 above)
Comment (tankangliang) [reply src/main/java/seedu/address/model/client/Client.java:31]: Nope, refer to PR description, tests currently rely on comparing two clients who are created at different timings, and this is pretty much just metadata for us I guess

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java (hunk 1 above)
Comment (qwoprocks) [reply src/main/java/seedu/address/storage/JsonAdaptedClient.java:159]: ok shld add a comment then, just to explain why its different from the rest

File: src/test/java/seedu/address/model/client/LastModifiedInstantTest.java (hunk 5 above)
Comment (tankangliang) [reply src/test/java/seedu/address/model/client/LastModifiedInstantTest.java:18]: Placed it together with the existing test.

**Review by qwoprocks — APPROVED** [review]
LGTM
